pkg/servicemesh/control-plane.go

```diff
@@ -0,0 +1,578 @@
+	if valid, _ := builder.validate(); !valid {
+		return builder
+	}
+
+	enablement := false
+
+	glog.V(100).Infof(
+		"Creating serviceMeshControlPlane %s with the all addons disabled", builder.Definition.Name)
+
+	if builder.Definition.Spec.Addons == nil {
+		builder.Definition.Spec.Addons = &istioV2.AddonsConfig{}
+	}
+
+	if builder.Definition.Spec.Addons.Grafana == nil {
+		builder.Definition.Spec.Addons.Grafana = &istioV2.GrafanaAddonConfig{}
+	}
+
+	if builder.Definition.Spec.Addons.Kiali == nil {
+		builder.Definition.Spec.Addons.Kiali = &istioV2.KialiAddonConfig{}
+	}
+
+	if builder.Definition.Spec.Addons.Jaeger == nil {
+		builder.Definition.Spec.Addons.Jaeger = &istioV2.JaegerAddonConfig{}
+	}
+
+	if builder.Definition.Spec.Addons.Prometheus == nil {
+		builder.Definition.Spec.Addons.Prometheus = &istioV2.PrometheusAddonConfig{}
+	}
+
+	if builder.Definition.Spec.Addons.Stackdriver == nil {
+		builder.Definition.Spec.Addons.Stackdriver = &istioV2.StackdriverAddonConfig{}
+	}
+
+	if builder.Definition.Spec.Addons.ThreeScale == nil {
+		builder.Definition.Spec.Addons.ThreeScale = &istioV2.ThreeScaleAddonConfig{}
+	}
+
+	disableAllAddons := &istioV2.AddonsConfig{
+		Prometheus: &istioV2.PrometheusAddonConfig{
+			Enablement: istioV2.Enablement{
+				Enabled: &enablement,
+			},
+		},
+		Jaeger: &istioV2.JaegerAddonConfig{
+			Install: &istioV2.JaegerInstallConfig{
+				Ingress: &istioV2.JaegerIngressConfig{
+					Enablement: istioV2.Enablement{
+						Enabled: &enablement,
+					},
+				},
+			},
+		},
+		Grafana: &istioV2.GrafanaAddonConfig{
+			Enablement: istioV2.Enablement{
+				Enabled: &enablement,
+			},
+		},
+		Kiali: &istioV2.KialiAddonConfig{
+			Enablement: istioV2.Enablement{
+				Enabled: &enablement,
+			},
+		},
+	}
+
+	builder.Definition.Spec.Addons = disableAllAddons
+
+	return builder
+}
+
+// WithGrafanaAddon adds grafana addons to the serviceMeshControlPlane.
+func (builder *ControlPlaneBuilder) WithGrafanaAddon(
+	enablement bool,
+	install *istioV2.GrafanaInstallConfig,
+	address *string) *ControlPlaneBuilder {
+	if valid, _ := builder.validate(); !valid {
+		return builder
+	}
+
+	glog.V(100).Infof(
+		"Creating serviceMeshControlPlane %s with the Grafana addons defined: enablement %s, install %s, "+
+			"address %s", builder.Definition.Name, enablement, install, address)
+
+	grafanaAddon := &istioV2.GrafanaAddonConfig{
+		Enablement: istioV2.Enablement{
+			Enabled: &enablement,
+		},
+		Install: install,
+		Address: address,
+	}
+
+	if builder.Definition.Spec.Addons == nil {
+		builder.Definition.Spec.Addons = &istioV2.AddonsConfig{}
+	}
+
+	if builder.Definition.Spec.Addons.Grafana == nil {
+		builder.Definition.Spec.Addons.Grafana = &istioV2.GrafanaAddonConfig{}
+	}
+
+	builder.Definition.Spec.Addons.Grafana = grafanaAddon
+
+	return builder
+}
+
+// WithJaegerAddon add joeger addons to the serviceMeshControlPlane.
+func (builder *ControlPlaneBuilder) WithJaegerAddon(
```

Comment:
Can you pass:
install *istioV2.JaegerInstallConfig here like you did in the WithGrafanaAddon function above?
It would reduce the number of arguments dramatically.  And the only 1 verification you'll need to run is:

```
	if jaegerAddonConfig == nil {
		glog.V(100).Infof("The jaegerAddonConfig of ServiceMeshControlPlane can not be empty")

		builder.errorMsg = "'jaegerAddonConfig' can not be empty for ServiceMeshControlPlane struct"
	}

```
You can built JaegerInstallConfig in your test cases for now and later we will implement jaegerinstallconfig struct in this package

Reply:
why?

Reply:
Due to following reasons: 
1. Function WithGrafanaAddon that was added by you above looks more readable.
2. Eliminates huge amount of arguments
3. Parameter Validation is simpler .
4. We don't need to pass all these arguments because most of them are part of more complex predefined JaegerInstallConfig structure.
5. Function is shorter and more readable 

Later we can add new file **jageraddonconfig.go** to this package with **JaegerAddonConfigBuilder**. We use the same approach for more complex packages like https://github.com/openshift-kni/eco-goinfra/tree/main/pkg/kmm
```
import (
	"fmt"
	"strings"

	"github.com/golang/glog"
	"github.com/openshift-kni/eco-goinfra/pkg/msg"
	istioV1 "maistra.io/api/core/v1"
	istioV2 "maistra.io/api/core/v2"
)

// JaegerAddonConfigBuilder builds JaegerAddonConfig struct based on given parameters.
type JaegerAddonConfigBuilder struct {
	// Module definition. Used to create a Module object.
	definition *istioV2.JaegerAddonConfig
	// Used in functions that define or mutate Module definition. errorMsg is processed before the Module
	// object is created.
	errorMsg string
}

// NewJaegerAddonConfigBuilder creates a new instance of NewJaegerAddonConfigBuilder.
func NewJaegerAddonConfigBuilder(name string) *JaegerAddonConfigBuilder {
	glog.V(100).Infof(
		"Initializing new JaegerAddonConfigBuilder structure with following params: %s", name)

	builder := &JaegerAddonConfigBuilder{
		definition: &istioV2.JaegerAddonConfig{
			Name: name,
		},
	}

	if name == "" {
		glog.V(100).Infof("The name of the JaegerAddonConfigBuilder is empty")

		builder.errorMsg = "'name' cannot be empty"
	}

	return builder
}

// WithStorage adds the specified InstallStorage config to the JaegerAddonConfigBuilder.
func (builder *JaegerAddonConfigBuilder) WithStorage(
	elasticsearchRedundancyPolicy string,
	memoryStorageMaxTraces int64,
	elasticSearchNodesCount int32,
	storageType *istioV2.JaegerStorageType,
	elasticsearchStorage *istioV1.HelmValues,
	elasticsearchIndexCleaner *istioV1.HelmValues,
) *JaegerAddonConfigBuilder {
	if valid, _ := builder.validate(); !valid {
		return builder
	}

	glog.V(100).Infof(
		"Creating JaegerAddonConfigBuilder structure with following storage parameters: "+
			"Type: %v, MaxTraces: %s, NodeCount: %s, Storage: %v, RedundancyPolicy: %s, IndexCleaner: %s",
		storageType, memoryStorageMaxTraces, elasticSearchNodesCount, elasticsearchStorage,
		elasticsearchRedundancyPolicy, elasticsearchIndexCleaner)

	if elasticsearchRedundancyPolicy == "" {
		glog.V(100).Infof("The elasticsearchRedundancyPolicy of JaegerAddonConfig is empty")

		builder.errorMsg = "'elasticsearchRedundancyPolicy' can not be empty for JaegerAddonConfig storage"
	}

	if memoryStorageMaxTraces < 1 {
		glog.V(100).Infof("The memoryStorageMaxTraces of JaegerAddonConfig is less than 1")

		builder.errorMsg = "'memoryStorageMaxTraces' can not be less than 1 for JaegerAddonConfig storage"
	}

	if elasticSearchNodesCount < 1 {
		glog.V(100).Infof("The elasticSearchNodesCount of JaegerAddonConfig is less than 1")

		builder.errorMsg = "'elasticSearchNodesCount' can not be less than 1 for JaegerAddonConfig storage"
	}

	if storageType == nil {
		glog.V(100).Infof("The storageType of JaegerAddonConfig can not be nil")

		builder.errorMsg = "'storageType' can not be nil for JaegerAddonConfig storage"
	}

	if elasticsearchStorage == nil {
		glog.V(100).Infof("The elasticsearchStorage of JaegerAddonConfig can not be nil")

		builder.errorMsg = "'elasticsearchStorage' can not be nil for JaegerAddonConfig storage"
	}

	if elasticsearchIndexCleaner == nil {
		glog.V(100).Infof("The elasticsearchIndexCleaner of JaegerAddonConfig can not be nil")

		builder.errorMsg = "'elasticsearchIndexCleaner' can not be nil for JaegerAddonConfig storage"
	}

	if builder.errorMsg != "" {
		return builder
	}

	if builder.definition.Install == nil {
		builder.definition.Install = &istioV2.JaegerInstallConfig{}
	}

	builder.definition.Install.Storage = &istioV2.JaegerStorageConfig{
		Type: *storageType,
		Memory: &istioV2.JaegerMemoryStorageConfig{
			MaxTraces: &memoryStorageMaxTraces,
		},
		Elasticsearch: &istioV2.JaegerElasticsearchStorageConfig{
			NodeCount:        &elasticSearchNodesCount,
			Storage:          elasticsearchStorage,
			RedundancyPolicy: elasticsearchRedundancyPolicy,
			IndexCleaner:     elasticsearchIndexCleaner,
		},
	}

	return builder
}

// WithIngress adds the specified InstallIngress config to the JaegerAddonConfigBuilder.
func (builder *JaegerAddonConfigBuilder) WithIngress(
	ingressEnablement bool,
	ingressLabels map[string]string,
	ingressAnnotations map[string]string) *JaegerAddonConfigBuilder {
	if valid, _ := builder.validate(); !valid {
		return builder
	}

	glog.V(100).Infof(
		"Creating JaegerAddonConfigBuilder structure with following Ingress parameters: "+
			"Enabled: %t, Labels: %v, Annotation: %v, Storage: %v, RedundancyPolicy: %s, IndexCleaner: %s",
		ingressEnablement, ingressLabels, ingressAnnotations)

	if len(ingressLabels) == 0 {
		glog.V(100).Infof("The ingressLabels map of JaegerAddonConfig is empty")

		builder.errorMsg = "'ingressLabels' can not be empty map for JaegerAddonConfig"
	}

	if len(ingressAnnotations) == 0 {
		glog.V(100).Infof("The ingressAnnotations map of JaegerAddonConfig is empty")

		builder.errorMsg = "'ingressAnnotations' can not be empty map for JaegerAddonConfig"
	}

	if builder.errorMsg != "" {
		return builder
	}

	if builder.definition.Install == nil {
		builder.definition.Install = &istioV2.JaegerInstallConfig{}
	}

	builder.definition.Install.Ingress = &istioV2.JaegerIngressConfig{
		Enablement: istioV2.Enablement{
			Enabled: &ingressEnablement,
		},
		Metadata: &istioV2.MetadataConfig{
			Labels:      ingressLabels,
			Annotations: ingressAnnotations,
		},
	}

	return builder
}

// GetJaegerAddonConfig returns JaegerAddonConfig with needed configuration.
func (builder *JaegerAddonConfigBuilder) GetJaegerAddonConfig() (*istioV2.JaegerAddonConfig, error) {
	if valid, err := builder.validate(); !valid {
		return nil, fmt.Errorf("error building JaegerAddonConfig config due to :%w", err)
	}

	return builder.definition, nil
}

// validate will check that the builder and builder definition are properly initialized before
// accessing any member fields.
func (builder *JaegerAddonConfigBuilder) validate() (bool, error) {
	resourceCRD := "DevicePluginContainer"

	if builder == nil {
		glog.V(100).Infof("The %s builder is uninitialized", strings.ToLower(resourceCRD))

		return false, fmt.Errorf("error: received nil %s builder", strings.ToLower(resourceCRD))
	}

	if builder.definition == nil {
		glog.V(100).Infof("The %s is undefined", strings.ToLower(resourceCRD))

		builder.errorMsg = msg.UndefinedCrdObjectErrString(resourceCRD)
	}

	if builder.errorMsg != "" {
		glog.V(100).Infof("The %s builder has error message: %s", strings.ToLower(resourceCRD), builder.errorMsg)

		return false, fmt.Errorf(builder.errorMsg)
	}

	return true, nil
}
```

Reply:
I can't entirely agree with your proposal; I did everything to make this object configuration more transparent and straightforward for future use. The complexity of this object configuration depends not on me and what I decided to do but on this object structure. And it is not clear in this case why we need infra. Just to create, delete, update, pull, and so on, a simple object view? and all other, more complex things we'll need to take care of on our tests level? I don't so understand, sorry.
Anyway, I removed it since there is only one way to progress with anything with an agreement with any of the reviewers.

Reply:
@elenagerman Please review my proposal above. By adding addition jageraddonconfig.go to or same structure for Prometheus and other addonConfig we do

1. Reduce complexity 
2. Keep same level of object configuration
3. Reduce code load from single structure 

Let's involve other maintainers @trewest @cdvultur @achuzhoy to this thread in order to have additional opinions.

Reply:
It makes sense to me that jaegerInstallConfig contain all of the configs required for setting up the addon rather than passing in the parameters individually. It reduces a large amount of validation burden on our end and follows the same logic as the other With*Addon methods.

Reply:
@kononovn @trewest @cdvultur @achuzhoy Could you please just review the last code version

Reply:
Definitely. I think that having an JaegerAddonConfigBuilder inside **jageraddonconfig.go** , based on the same example from [KMM](https://github.com/openshift-kni/eco-goinfra/tree/main/pkg/kmm) would look better.

With that you can control better the JaegerAddonConfig, allowing extra enhancements in the future. And those enhancements will not affect the ControlPlaneBuilder.